DOCSP-9563 crud search text (#85)
mongodb/docs-java: source/fundamentals/crud/read-operations/text.txt

   { "_id": 2, "title": "Fast 5", "tags": ["bank robbery", "full team"] }
   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

Text Index
~~~~~~~~~~

You must create a **text index** before running a text search. A text
index specifies the string or string array field on which to run a text
search. For more information, see the :ref:`Text Indexes <textIndexes>`
section of our Indexes guide.

In the following examples, you run text searches on the ``title``
field in the ``fast_and_furious_movies`` collection. To enable text
searches on the ``title`` field, create a text index using the
:ref:`Indexes <index-text-indexes>` builder with the following 
snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin textIndex
   :end-before: end textIndex

Text Search
-----------

Use the ``Filters.text()`` method to specify a text search. 

The ``Filters.text()`` method uses the :doc:`Filters builder
</fundamentals/builders/filters>` to define a query filter specifying

> what to search for during the text search. You represent the query filter

Reviewer: When "you represent", it generally means that you speak for someone or that something amounts to something. I think in this case, the query filter is represented by Bson rather than the reader speaking for the query filter or amounting to the query filter. Here's my previous suggestion:
"The query filter is represented by a Bson instance."
Link to the full suggestion: https://github.com/mongodb/docs-java/pull/85/files/47222d03539ffe913c8b47637e2a7fbcba7efcc1#r647027059

   :end-before: end multipleTermExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

Search Text by a Phrase
~~~~~~~~~~~~~~~~~~~~~~~

Pass a phrase with escaped quotes to the ``Filters.text()`` method to
specify the phrase in your text search. If you don't add escaped quotes,
the method runs a :ref:`term search <term_search>`. 

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
phrase "Fate of the Furious". 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin phraseExample
   :end-before: end phraseExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }
 
Search Text with Terms Excluded
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For each term you want to exclude from your text search, prefix the term

> with a minus sign in string you pass to the ``Filters.text()`` builder

Reviewer: Sorry, my prior github suggestion comment that you used had a grammatical error in it. Can you fix it for me here:
"with a minus sign in the string that you pass..."

method. 

You must have at least one text search term if you want to
exclude terms from your search. If you have multiple text indexes, the

> search applies to all indexes. 

Reviewer: From reading your previous revisions, I thought you can only have one text index. I think the only exception is if you use the Atlas full text search feature which we haven't covered in this document.

Author: In the text search section, I have 
> When you execute the ``find()`` method, MongoDB runs a text search on
the fields indexed with the text index on the collection.

I think this covers multiple text indexes as well?

Reviewer: I think my comment referred to an older commit:

> If you have multiple text indexes, the
> search applies to all indexes.

Each document matched does not contain the excluded term in
your text index field. 

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the

> term "Furious", but do not contain the term "Fast".

Reviewer: Nitpick: Similar comment https://github.com/mongodb/docs-java/pull/85/files/a682aaaa0f606d5516304e16e46eb0b9b0e28b00#r646777048

   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

Text Index
~~~~~~~~~~

You must create a **text index** before running a text search. A text
index specifies the string or string array field on which to run a text
search.

In the following examples, you run text searches on the ``title``
field in the ``fast_and_furious_movies`` collection. To enable text
searches on the ``title`` field, create a text index using the
:ref:`Indexes <index-text-indexes>` builder with the following 
snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin textIndex
   :end-before: end textIndex

For more information, see the :ref:`Text Indexes <textIndexes>` section
of our Indexes guide. 

Text Search
-----------

Use the ``Filters.text()`` method to specify a text search. 

The ``Filters.text()`` method uses the :doc:`Filters builder
</fundamentals/builders/filters>` to define a query filter specifying
what to search for during the text search. The query filter is

> represented by :ref:`Bson <bson>` instance. Pass the query filter to the

Reviewer: ```suggestion
represented by a :ref:`Bson <bson>` instance. Pass the query filter to the
```

:ref:`relevance score for each result <sorts-crud-text-search>`. 

Specify Options
~~~~~~~~~~~~~~~

You can include ``TextSearchOptions`` as the second parameter of the
``Filters.text()`` method to specify text search options such as case
sensitivity. By default, text searches run without case sensitivity
which means the search matches lowercase and uppercase values.

To specify a case sensitive search, use the following snippet:

.. code-block:: java

   TextSearchOptions options = new TextSearchOptions().caseSensitive(true);
   Bson filter = Filters.text("SomeText", options);

For more information on the method and class in this section, see the following API documentation:

- :java-core-api:`Filters.text() <com/mongodb/client/model/Filters.html#text(java.lang.String)>`
- :java-core-api:`TextSearchOptions <com/mongodb/client/model/TextSearchOptions.html>`

.. _term_search:

Search Text by a Term
~~~~~~~~~~~~~~~~~~~~~

Pass a term as a string to the ``Filters.text()`` method to specify the
term in your text search. 

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the

> term "fast".

Reviewer: Nitpick: Before a code snippet, should the sentence end with ":" or ":"? And is there a difference between the two? If so, what's the difference? Or, should they be consistent as one or the other? 
Applies to all instances.

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin termExample
   :end-before: end termExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 1, "title": "2 Fast 2 Furious ", "tags": ["undercover", "drug dealer"] }
   { "_id": 2, "title": "Fast 5", "tags": ["bank robbery", "full team"] }

To match multiple terms in your text search, separate each term

> with spaces in the ``Filters.text()`` builder method. When you pass the 
> text search query returned by the builder method to the
> ``find()`` method, it returns documents that match any of the terms. 

Reviewer: I think there might be too many different cause-effect items in this sentence which makes it confusing to read. While slightly arbitrary, the style guide mentions that we should try to use a sentence limit of 26 words: https://docs.mongodb.com/meta/style-guide/writing/write-for-accessibility/#write-for-ease-of-reading

Consider splitting the sentence to only contain one cause and one effect. Currently, there's a chain of them:

"When you pass N returned by M to O, it returns P."

Here's one way you can accomplish this
```suggestion
with spaces in the ``Filters.text()`` builder method. The builder method
returns the text search query as a ``Bson`` instance. When you pass
this to the ``find()`` method, it returns documents that match any of the terms.
```

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
terms "fate" or "7".

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin multipleTermExample
   :end-before: end multipleTermExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

Search Text by a Phrase
~~~~~~~~~~~~~~~~~~~~~~~

> Pass a phrase with escaped quotes to the ``Filters.text()`` method to

Reviewer: From a Jasmin persona reader, I wonder whether "escaped quotes" might be jargon which should be defined/demonstrated as you might do for a term. E.g.
Escaped quotes are double quote characters preceded by a backslash character.

And then specify that the positioning of the escaped quotes.
"If you don't add escaped quotes around the phrase, the method performs..."

specify the phrase in your text search. If you don't add escaped quotes,
the method runs a :ref:`term search <term_search>`. 

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
phrase "fate of the furious". 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin phraseExample
   :end-before: end phraseExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }
 
Search Text with Terms Excluded
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For each term you want to exclude from your text search, prefix the term
with a minus sign in the string that you pass to the ``Filters.text()``
builder method. 

> Each document matched does not contain the excluded term in
> your text index field. 

Reviewer: Emphasize none to specify "any" rather than individually ("each does not").
```suggestion
None of the documents returned from the search contain the excluded term in
your text index field. 
```

.. note::

   You must have at least one text search term if you want to
   exclude terms from your search. 

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
term "furious", but do not contain the term "fast".

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin negateExample
   :end-before: end negateExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

For more information on text indexes, see our Server Manual Entry on :manual:`Text

> Indexes </core/index-text/>`. 

Reviewer: I'm not sure it has anything to do specifically with the example of excluding a term. Wouldn't it make more sense to put in the Text Indexes section of this page next to the link to the Indexes guide?

Author: It does, but the Indexes guide is linked in the Text Index section -- so should the same link be listed twice?

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you will learn how to run a **text search** in MongoDB.

You can use a text search to retrieve documents that contain a **term**
or a **phrase** in a specified field. A term is a sequence of characters
that excludes whitespace characters. A phrase is a sequence of terms
with any number of whitespace characters. 

The following sections show you how to perform the following types of
text searches:

- Search Text by a Term
- Search Text by a Phrase
- Search Text with Terms Excluded

If you want to sort your text search results, see the :ref:`Text Search
<sorts-crud-text-search>` section of our Sort Results guide. 

Sample Documents
~~~~~~~~~~~~~~~~

The following sections feature examples of text searches on the
``fast_and_furious_movies`` collection. Each section uses a variable
named ``collection`` to refer to the ``MongoCollection`` instance of the
``fast_and_furious_movies`` collection. 

The ``fast_and_furious_movies`` collection contains documents that
describe one of the several movies that are part of the Fast and Furious

> move franchise. Each document contains a title field and a tags field.

Reviewer: ```suggestion
movie franchise. Each document contains a title field and a tags field.
```

To match multiple terms in your text search, separate each term
with spaces in the ``Filters.text()`` builder method. The builder method
returns the text search query as a ``Bson`` instance. When you pass
this to the ``find()`` method, it returns documents that match any of
the terms.

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
terms "fate" or "7":

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin multipleTermExample
   :end-before: end multipleTermExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

Search Text by a Phrase
~~~~~~~~~~~~~~~~~~~~~~~

Pass a phrase with **escaped quotes** to the ``Filters.text()`` method to
specify the phrase in your text search. Escaped quotes are double quote
characters preceded by a backslash character. If you don't add escaped

> quotes around the phrase, the method runs a :ref:`term search

Reviewer: If "the method" is "Filters.text()", it specifies the search and doesn't run it right? If you're referring to find(), you'll need to specify that somewhere.

===========
Search Text
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you will learn how to perform a text search in MongoDB.

Performing a text search allows you to search a string field in your
collection for terms or phrases. This is useful for searching
unstructured text such as transcripts, resumes, essays, or web pages. 

Sample Documents
~~~~~~~~~~~~~~~~

> In the following examples, a junior year student is browsing through

Reviewer: Not sure whether you've addressed my comment here. While we use the term "junior" in the US to generally describe the third or second-last year of college, it's not referred to as such in other countries like the UK. Does "junior" add anything to the understanding of the example or are you using it to contrast between something else?

string or an array of string elements. 

In the following examples, we perform text searches on the ``course``
field in the ``junior_courses`` collection. To enable text searches on
the ``course`` field, create a text index using the following snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin textIndex
   :end-before: end textIndex

If there was another string field, ``description``, you wanted to index,
you can create a compound index to perform a text search in both fields
using the following snippet: 

.. code-block:: java

    collection.createIndex(Indexes.compoundIndex(Indexes.text("course"), Indexes.text("description")));

For more information, see our server manual entry on :manual:`text
indexes </core/index-text/>`.

Searching
---------

To perform a text search, use the ``text()`` method of the ``Filters``
builder. 

The ``text()`` method performs a text search on the fields indexed with
a text index. The text search matches documents which contain the
specified search term. 

You can include ``TextSearchOptions`` as the second parameter to specify

> additional text search options to handle case sensitivity, diacritic

Reviewer: I think case sensitivity is probably a common use case. I think it would be great if you could mention whether searches are case sensitive by default. And it would be wonderful to have an example in which you're setting it to whatever is not the default.

sensitivity and the language of the text. 

For more information, see the API Documentation on the
:java-core-api:`text()
<com/mongodb/client/model/Filters.html#text(java.lang.String)>` method
and :java-core-api:`TextSearchOptions<com/mongodb/client/model/TextSearchOptions.html>`.

By a Term
~~~~~~~~~

To search for a term in your text index, pass the term as a string
to the ``text()`` method. 

If you want to match multiple terms in your text index, separate your
terms with spaces. The ``text()`` method performs a logical OR on
each term and returns documents that contain any of the terms. 

Example 
```````

The student would like to see which courses mention "Computer" in their
name. 

To address this scenario, the student performs a text search on

> documents that contain the term "Computer": 

Author: I am choosing to not mention it because there's a section on the page dedicated to talking about creating a text index.

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin wordExample
   :end-before: end wordExample

> The following shows the output of the code above:

Author: https://docs.mongodb.com/meta/style-guide/style/links-and-cross-references/

I've fixed each instance by using "preceding".

> The following shows the output of the code above:

Reviewer: Great! You meant https://docs.mongodb.com/meta/style-guide/terminology/general-term-guidelines/use-consistent-time-references/ right?

.. code-block:: json
   :copyable: false

   { "_id": 2, "course": "Discrete Mathematical Structures for Computer Science" }
   { "_id": 4, "course": "Computer Science" }
   { "_id": 5, "course": "Computer Hacking Revealed" }
   { "_id": 6, "course": "Advanced Computer Game Programming" }
   { "_id": 7, "course": "Image Processing in Computer Science" }

.. note::

    The output of the snippet above is equivalent to performing the
    following text search:

    .. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
       :language: java
       :dedent:
       :start-after: begin multiWordExample
       :end-before: end multiWordExample

By a Phrase
~~~~~~~~~~~

To search for a phrase in your text index, pass the phrase with escaped quotes
to the ``text()`` method.  

If you don't add escaped quotes, the method performs a logical OR on the

> terms listed.

Reviewer: It might make more sense to the reader to mention that it performs a term search without the escaped quotes. It's a little easier to read than an explanation of what the term search does.

Example 
```````

The student would like to know which courses mention "Computer Science"
in their name. 

To address this scenario, the student performs a text search on
documents that contain the phrase "Computer Science":

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin phraseExample
   :end-before: end phraseExample

The following shows the output of the code above:

.. code-block:: json
   :copyable: false

   { "_id": 2, "course": "Discrete Mathematical Structures for Computer Science" }
   { "_id": 4, "course": "Computer Science" }
   { "_id": 7, "course": "Image Processing in Computer Science" }

Excluding Terms
~~~~~~~~~~~~~~~

To exclude a term in your text search, prefix a minus sign, ``-``, to
each term you want to omit. 

Example 
```````
The student would like to know which courses mention "Computer" without
"Science".

To address this scenario, the student performs a text search on

> documents that contain the term "Computer", but not contain the term

Reviewer: ```suggestion
documents that contain the term "Computer", but do not contain the term
```

"Science": 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin negateExample
   :end-before: end negateExample

The following shows the output of the code above:

.. code-block:: json
   :copyable: false

   { "_id": 5, "course": "Computer Hacking Revealed" }
   { "_id": 6, "course": "Advanced Computer Game Programming" }

Sort by Relevance
~~~~~~~~~~~~~~~~~

You can see how relevant the documents returned by the text are by
including a projection to calculate the ``metaTextScore``. The
``metaTextScore`` signifies how well the document matched the search
term.

Example 
```````
The student would like to see which course mentioning "Computer Science"
is the most relevant. 

To address this scenario, the student performs the following actions:

- A text search for the term "Computer Science"
- A projection to calculate the ``metaTextScore`` and place the score in a field called ``score``

> - A sort to display the results in descending order

Reviewer: Remember that we shouldn't assume our readers know about any of the methods or classes because they could have arrived at the documentation from anywhere. We should reference Projections and Sorts builders somewhere, ideally before we talk about projections and sorts. We should also add API documentation links at the end of this section.

===========
Search Text
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you will learn how to perform a **text search** in MongoDB.

> A text search allows you to search for partial or exact matches from a set
> of words. You can use this for searching through unstructured text, which is

Reviewer: I'm not sure partial or exact matches describes what it's matching. Instead, how about the following:

```suggestion
You can use a text search to search documents for terms or phrases. A term is a sequence of characters that excludes whitespace characters. A phrase is a sequence of characters that may include any number of whitespace characters. You can use this for searching through unstructured text, which is
```

> text without a predefined format such as resumes and websites. 

Author: Great point, I'll change it to things that don't overlap as much.

In MongoDB, you can perform a text search on a string field in your
collection. 

Sample Documents
~~~~~~~~~~~~~~~~

> In the following examples, a student is browsing through courses to

Author: I couldn't seem to find a good way to mention it here, so I mentioned it as an example of unstructured text above.

enroll in at their college. The documents in the ``junior_courses``
collection represents the courses available with their intended majors: 

.. code-block:: json

    { "_id": 1, "name": "Forensic Science", "major": ["Forensic Science", "Criminal Justice"] }
    { "_id": 2, "name": "Discrete Mathematical Structures for Computer Science", "major": ["Computer Science", "Mathematics"] }
    { "_id": 3, "name": "Environmental Science", "major": ["Environmental Science", "Sustainable Management"] }

>     { "_id": 4, "name": "Image Processing in Computer Science", "major": ["Artificial Intelligence", "Computer Science"] }

Author: Apologies, I forgot to update the documents here to reflect what I changed them to.

Text Index
----------

Create a **text index** using the ``Indexes`` builder before you perform
a text search. A text index specifies the string or string array field
to query during a text search. 

> 

Reviewer: I still think it would be good to explain why and that it doesn't apply to Atlas.

> 

Reviewer: Bump ^

> 

Reviewer: Atlas info: https://docs.mongodb.com/manual/core/index-text/#create-text-index

> 

Author: I'm removing this section and redirecting it to the Indexes page since it's covered there.

In the following examples, the student performs text searches on the ``name``
field in the ``junior_courses`` collection. To enable text searches on
the ``name`` field, the student creates a text index using the following
snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin textIndex
   :end-before: end textIndex

If the student wanted to index both the ``name`` and ``major`` field,

> the student can modify the previous index to a compound index. A

Author: I meant just a text index that can index multiple fields.

compound index is a single index structure holds references to multiple
fields within a collection's documents. To enable text searches on
the ``name`` and ``major`` field, the student creates a compound index

> using the following snippet: 

Author: I'll reword it to portray an alternative way to create an index. The compound index was more of an alternative way if the user wanted to index multiple fields.

.. code-block:: java

    collection.createIndex(Indexes.compoundIndex(Indexes.text("name"), Indexes.text("major")));

.. note::

   The student cannot create a second index because there can only be one
   text index per collection. 

For more information, see the following documentation:

- :manual:`Text Indexes </core/index-text/>` Server Manual Entry
- :ref:`Text Indexes <index-text-indexes>` in our Indexes builder guide
- :ref:`Compound Indexes <index-compound-indexes>` in our Indexes builder guide

Searching
---------

Use the ``Filters.text()`` method to perform a text search. 

The ``Filters.text()`` method uses the :doc:`Filters builder
</fundamentals/builders/filters>` to perform the text search on the
fields indexed with the text index on the collection. The text search
matches documents which contain one or more search terms.

You can include ``TextSearchOptions`` as the second parameter to specify
text search options such as case sensitivity. By default, there is no
case sensitivity, meaning MongoDB matches documents regardless if the
value is lowercase or uppercase. 

> 

Reviewer: I think the items: "certain terms" and "how to sort the relevance of your text search" are not very clear. Maybe it would be helpful just to list them by the names of the sections? Also, not sure the reader knows how to distinguish between a section and subsection. E.g.
The following sections show you how to perform the following types of text searches:
- Search Text by a Term
- Search Text by a Phrase
- Search Text with Terms Excluded
- Sort by Relevance

each term and returns documents that contain any of the terms. 

Example 
```````

The student wants to find Science courses with case sensitivity on
the first letter being capital.

To address this scenario, the student performs the following:

- Specify to have the search be case sensitive though ``TextSearchOptions``
- Perform a text search on documents that contain the term "Science" with the ``Filters.text()`` method

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin caseSensitiveExample
   :end-before: end caseSensitiveExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 2, "name": "Discrete Mathematical Structures for Computer Science", "major": ["Computer Science", "Mathematics"] }
   { "_id": 3, "name": "Environmental Science", "major": ["Environmental Science", "Sustainable Management"] }

.. note::

   The output would consist of all the documents in the
   ``junior_courses`` collection if the search was not case sensitive
   because each document contains the term "Science" in the ``name``

>    field. 

Reviewer: The order of information seems reversed. How about ordering the cause before the effect, e.g.:
"If the search was not case sensitive... the output would..."

Search Text By a Phrase
~~~~~~~~~~~~~~~~~~~~~~~

Pass the phrase with escaped quotes to the ``Filters.text()`` method to
search for a phrase in your text search. If you don't add escaped quotes,

> the method performs a :ref:`term search <term_search>`. 

Author: Does defining it in the overview satisfy this?

Example 
```````

The student wants to find Computer Science courses.

To address this scenario, the student performs a text search on
documents that contain the phrase "Computer Science":

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin phraseExample
   :end-before: end phraseExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 2, "name": "Discrete Mathematical Structures for Computer Science", "major": ["Computer Science", "Mathematics"], "score": 1.2 }

>    { "_id": 4, "name": "Image Processing in Computer science", "major": ["Artificial Intelligence", "Computer Science"], "score": 1.25 }

Author: I corrected the sample data at the top

 
Search Text With Terms Excluded
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Prefix a minus sign ``-`` to each term you want to exclude in your text

Author: Yes, that would be the syntax.

search. 

Example 
```````
The student wants to only find Science courses.

To address this scenario, the student performs a text search on

> documents that contain the term "Science", but do not contain the term

Reviewer: I would make sure it's clear that you're searching on a collection for documents that contain some value in a field.

"Computer": 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin negateExample
   :end-before: end negateExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

>    { "_id": 1, "name": "Forensic science", "major": ["Forensic Science", "Criminal Justice"] }

Author: I forgot to update the sample data above -- document 1 and 4 are supposed to have a lowercase `s` in science

   { "_id": 3, "name": "Environmental Science", "major": ["Environmental Science", "Sustainable Management"] }
  
Sorting by Relevance
~~~~~~~~~~~~~~~~~~~~

You can retrieve a score to see how relevant the returned documents are

> in your text search. To retrieve a score, project the ``metaTextScore``

Reviewer: "metaTextScore" might be a new term for readers. I wonder if it would be helpful to describe in this section.

in your query.

Example 
```````

> The student wants to find Computer Science courses and view the courses

Reviewer: You can use the pronoun "them" to refer to "the courses" since they're the last object prior to that in the sentence.

in order of relevance to their search.

To address this scenario, the student performs the following:

> - Perform a text search for the phrase "Computer Science"

Reviewer: There's a switch in "performs" to "perform". Is this intentional, and if so, does it sound correct? Or does sticking to either "performs" or "perform" work better?

Author: I'm hesitant to change it to performs because each bullet doesn't start with a word in the `'s` form.

- Specify a :doc:`Projections builder <fundamentals/builders/projections>` to calculate the ``metaTextScore`` and place the score in a field called ``score``

> - Specify a :doc:`Sorts builder <fundamentals/builders/sort>` to display the results in descending order

Reviewer: Also, it could help to explain why it's descending order. Is that by default? I wasn't able to find that info in this PR.

===========
Search Text
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you will learn how to run a **text search** in MongoDB.

You can use a text search to retrieve documents that contain a **term**
or a **phrase** in a specified field. A term is a sequence of characters
that excludes whitespace characters. A phrase is a sequence of terms
that with any number of whitespace characters. 

The following sections show you how to perform the following types of
text searches:

- Search Text by a Term
- Search Text by a Phrase
- Search Text with Terms Excluded

> If you want to sort by relevance, see our :ref:`Sort Results
> <sorts-crud-text-search>` page.

Reviewer: ```suggestion
Not sure if a reader will know what relevance is, so maybe omitting that and mentioning that you're sorting results of a text search would be more clear.

```suggestion
If you want to sort your text search results, see the :ref:`Text Search <sorts-crud-text-search>` section of our Sort Results guide.
```

``fast_and_furious_movies`` collection. 

The collection contains documents with a title and tags associated
with one of the several movies that are part of the Fast and Furious
movie franchise: 

.. code-block:: json

   { "_id": 1, "title": "2 Fast 2 Furious ", "tags": ["undercover", "drug dealer"] } 
   { "_id": 2, "title": "Fast 5", "tags": ["bank robbery", "full team"] }
   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

Text Index
----------

You must create a **text index** before running a text search. A text
index specifies the string or string array field on which to run a text
search. A collection can only contain one text index.

In the following examples, you run text searches on the ``title``
field in the ``fast_and_furious_movies`` collection. To enable text
searches on the ``title`` field, create a text index using the
:ref:`Indexes <index-text-indexes>` builder with the following 
snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin textIndex
   :end-before: end textIndex

> Multiple Indexes

Reviewer: I think "Multiple Indexes" might be slightly misleading since you're explaining here that you need to use a compound index instead of more than one (multiple) index.
Multiple Indexed Text Fields or Compound Text Index might be better.

Author: I'm removing this section and redirecting to the Indexes page.

~~~~~~~~~~~~~~~~

If you want to run text searches on both the ``title`` and
``tags`` field in the ``fast_and_furious_movies`` collection, you
need to create an text index to include them.

You can create a :ref:`compound index <index-compound-indexes>`. A
compound index is a single index that refers to multiple
fields within a collection's documents. This allows you to specify
multiple text fields using one text index.

To enable text searches on the ``title`` and ``tag`` fields, create a
compound index using the following snippet:  

.. code-block:: java

   collection.createIndex(Indexes.compoundIndex(Indexes.text("title"), Indexes.text("tags")));

.. note::

>    If you have an existing text index on the collection, you need to drop
>    it. You can drop an index by calling the ``dropIndex()`` method,

Reviewer: Currently sounds like this is asking to drop any existing text index regardless of what you're trying to do.

```suggestion
   Since you can only have one text index on your collection, you need to drop 
   any existing text indexes to create a new one.
   You can drop an index by calling the ``dropIndex()`` method,
```

Reviewer: Where can the reader find out how to drop an index or create a compound index? I thought that could be important. If it's located somewhere else, can you link to it in this document?

Reviewer: Compound text indexes have a few restrictions that I don't think are listed on that page. For more info on those restrictions:
https://docs.mongodb.com/manual/core/index-text/#std-label-text-index-compound

I think what you had in the section was good info, just needed some additional info.

Author: Instead of bringing over existing content, I linked to that manual page.

   specifying name of the text index as shown in the following snippet:

   .. code-block:: java

      collection.dropIndex("field_name");
   
   If you want to delete all existing indexes, specify "*" instead of

>    the name. Note that this does not delete the required index on the

Reviewer: Is this a common case? If so, I wonder whether this information belongs on the Indexes page instead.

Author: Removed this section.

   ``_id`` field. 

Text Search
-----------

Use the ``Filters.text()`` method to specify a text search. 

> The ``Filters.text()`` method uses the :doc:`Filters builder
> </fundamentals/builders/filters>` to create an instance of a :ref:`Bson
> <bson>` type. The instance is a query filter that specifies how the text
> search should operate, which you pass to the ``find()`` method. 

Reviewer: Thought I had created a comment about this, but maybe I removed it by accident.
We should introduce what the Bson represents before the underlying type. A "Bson instance" could be anything you pass to MongoDB server, so we want to qualify it first. How about this:

```suggestion
The ``Filters.text()`` method uses the :doc:`Filters builder
</fundamentals/builders/filters>` to define a query filter based on
the search that you pass it. The query filter is represented by a 
Bson instance. Pass the query filter to the ``find()`` method to 
run your text search.
```

   :language: java
   :dedent:
   :start-after: begin multipleTermExample
   :end-before: end multipleTermExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

Search Text by a Phrase
~~~~~~~~~~~~~~~~~~~~~~~

Pass a phrase with escaped quotes to the ``Filters.text()`` method to
specify the phrase in your text search. If you don't add escaped quotes,
the method runs a :ref:`term search <term_search>`. 

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
phrase "Fate of the Furious". 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin phraseExample
   :end-before: end phraseExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }
 
Search Text with Terms Excluded
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Pass a term prefixed with a minus sign ``-`` to the ``Filters.text()``
> method to specify each term you want to exclude in your text search.

Reviewer: A couple grammar issues. First, there needs to be singular form agreement (currently, you're referring to one in the subject and an indeterminate number in "to specify each term" phrase).  How about this:

```suggestion
For each term you want to exclude from your text search, prefix the term with a minus sign in string you pass to the Filters.text() builder method.
```

> You must have at least one term to include to specify terms to exclude.

Reviewer: Not sure what "at least one term to include" means. Perhaps you can reword as follows:
```suggestion
You must have at least one text search term if you want to exclude terms from your search.
```

> If you have multiple text indexes, the search applies to all indexes.

Reviewer: A couple items here:
1) Does this sentence have anything to do with excluding terms? If you're addressing [this comment](https://github.com/mongodb/docs-java/pull/85/files#r646784453), I was looking for info on whether the exclusion would apply to all indexed fields.
2) A text search doesn't "[apply] to indexes". Instead, a search can use indexed data to retrieve results. Instead of the current sentence, consider the following:

```suggestion
If your collection has more than one text field indexed, the search and exclusion applies to all the indexed fields.
```

Author: Removed this section.

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
term "Furious", but do not contain the term "Fast".

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin negateExample
   :end-before: end negateExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

> For more information, see the server manual entry on :manual:`Text

Reviewer: For consistency, should "server manual entry" be capitalized?

string or string array field for the query to run a text search. A
collection can only contain one index.

In the following examples, we run text searches on the ``title``
field in the ``fast_and_furious_movies`` collection. To enable text searches on
the ``title`` field, create a text index using the following
snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin textIndex
   :end-before: end textIndex

If we wanted to run text searches on both the ``title`` and
``tags`` field in the ``fast_and_furious_movies`` collection, we need to
drop the index for the ``title`` field with the following snippet:

.. code-block:: java

   collection.dropIndex("title");

Then, we can create a compound index. A compound index is a
single index structure holds references to multiple fields within a
collection's documents. This allows us to specify multiple text indexes
without violating the constraint for a collection to contain one 
index. 

To enable text searches on the ``title`` and ``tag`` field, create a
compound index using the following snippet: 

.. code-block:: java

>    collection.createIndex(Indexes.compoundIndex(Indexes.text("title"), Indexes.text("tags")));

Author: I'll put this alternative in a note box

   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }
 
Search Text with Terms Excluded
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Prefix each term you want to exclude in your text search with a minus
sign ``-``. 

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
term "Furious", but do not contain the term "Fast".

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin negateExample
   :end-before: end negateExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

.. _sort_by_relevance:

Sort by Relevance
~~~~~~~~~~~~~~~~~

The text search assigns a numerical text score to indicate how relevant
each result matches the search string. Often, you want to know which
results are the most relevant.  

To sort by relevance, you use the :doc:`Projections
</fundamentals/builders/projections>`  and :doc:`Sorts

> </fundamentals/builders/sort>` builders.

Reviewer: From the paragraphs below, I see you understand what the builders are for, but I think item 1 still hasn't been addressed.

When you start the sentence with "To sort by relevance", the remainder of the sentence should answer what operations must you perform to accomplish that. You don't need the Projects or Sorts builders to accomplish this.

Instead, let's try this:

1) Start the paragraph with "To perform a text search and sort the results by text score relevance, you must specify the score field as parameters in both your sort and projection methods."
This is describing how they need to perform collection.find().projection().sort() without mentioning the helpers.
2) Now, we talk about the helpers that you can use to specify the parameters to those methods. "You can use the Projection.metaTextScore() builder method to specify the projection and the Sorts.metaTextScore() builder method to specify the sort."
3) The above properly introduces the two new paragraphs you created to describe the builders.

If this doesn't make complete sense, let's discuss this over Zoom because it's an important distinction.

First, create a project document to retrieve a text score by using the
``Projections.metaTextScore()`` method. Upon passing this document to the
``projection()`` method generates a new field with the specified name
and returns the documents text score. 

Then, specify a sort criteria to sort by the text score by using the
``Sorts.metaTextScore()`` method. Upon passing this to the ``sort()``
method, MongoDB sorts the results from the most to least relevant. By

> default, the method returns the results unsorted. 

Reviewer: It would be helpful to mention what you're calling the sort() method on.

Reviewer: I meant that you call sort() on an instance of a MongoCollection. Since the reader might not know much about sort() yet as they're on the Text Search page, we should make this clear in a location that a reader might expect it such as when it's first introduced.

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the term
"Fast". 

> To view them in order of relevance to the search, perform the following
> actions: 

Reviewer: ```suggestion
To view the results in order of relevance to the search term, perform the following
actions: 
```

- Project the text scores of the query results as the ``score`` field

> - Sort the results by text score (most relevant match first)

Reviewer: I think these are full sentences, not fragments, so we should add "." at the end.

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin scoreExample
   :end-before: end scoreExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 2, "title": "Fast 5", "tags": ["bank robbery", "full team"], "score": 0.75 }
   { "_id": 1, "title": "2 Fast 2 Furious ", "tags": ["undercover", "drug dealer"], "score": 0.625 }
   

> The movie with the ``_id`` value of 2 is the most relevant to our
> search. Although both documents contain what the search phrase we

Reviewer: I think it could be better to talk about the document by its title in this case.
```suggestion
The movie "Fast 5" is the most relevant result in our search. Although both documents contain the search phrase we
```

> were looking for, MongoDB then ranks based on the string length of the

Reviewer: Maintain tense (in this case, past):

```suggestion
were looking for, MongoDB ranked them based on the string length of the
```

I actually prefer present tense, but past tense is not a blocker.

``title`` field. 

For more information, see the following documentation:

- :java-docs:`Projections.metaTextScore() <apidocs/mongodb-driver-core/com/mongodb/client/model/Projections.html#metaTextScore(java.lang.String)>` API documentation
- :java-docs:`Sorts.metaTextScore() <apidocs/mongodb-driver-core/com/mongodb/client/model/Sorts.html#metaTextScore(java.lang.String)>` API documentation
- :manual:`Text score metadata </reference/operator/aggregation/meta/#mongodb-expression-exp.-meta>` Server Manual Entry 

> - :manual:`Text search score </reference/operator/query/text/#return-the-text-search-score/>` Server Manual Entry 

Reviewer: Shouldn't these be title case if you're using title case for "Server Manual Entry"?

===========
Search Text
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you will learn how to run a **text search** in MongoDB.

You can use a text search to search documents for a term or a phrase. A
term is a sequence of characters that excludes whitespace characters. A
phrase is a sequence of terms that with any number of whitespace
characters. You can use this for searching through unstructured text,
which is text without a predefined format such as resumes and

> course catalogs. 

Reviewer: Perhaps it's the way it is worded, but this makes me wonder what is considered structured text and why you can't search it with text search. Could you either explain it in more detail or omit the term if possible to avoid confusion?

In MongoDB, you can run a text search on a string field. A text
search retrieves documents that contain the term or phrase in the

> specified field. 

Reviewer: I would combine this with the previous paragraph because the text search you're describing there is not a general one, but rather specific to MongoDB.
E.g.
"In MongoDB, a text search retrieves documents that contain a matching term or phrase in one or more string fields."

Sample Documents
~~~~~~~~~~~~~~~~

In the following examples, a student browses through courses to
enroll in at their college. The documents in the ``junior_courses``
collection represent the courses available for junior year students
with their intended majors: 

.. code-block:: json

    { "_id": 1, "name": "Forensic science", "major": ["Forensic Science", "Criminal Justice"] }
    { "_id": 2, "name": "Discrete Math in Computer Science", "major": ["Computer Science", "Mathematics"] }
    { "_id": 3, "name": "Environmental Science", "major": ["Environmental Science", "Sustainable Management"] }
    { "_id": 4, "name": "Computer science Seminar", "major": ["Computer Science"] }

Text Index
----------

> You must create a **text index** using the ``Indexes`` builder before

Reviewer: This sentence is worded such that using the Index builder is necessary, but it's not. You can create a text index using anything that converts to a BSON document. Instead, separate the "you must" to apply only to the text index and apply "you can" to the Indexes builder (as an optional way to create it).

running a text search. A text index specifies the string or string
array field for the query to run a text search. There can only be

> one text index per collection. 

Reviewer: I think "there can be..." is passive voice. Maybe "A collection can only contain one text index." would be more clear?

the ``name`` field, the student creates a text index using the following
snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin textIndex
   :end-before: end textIndex

If the student wanted to run text searches on both the ``name`` and
``major`` field in the ``junior_courses`` collection, they can instead
create a compound index. A compound index is a single index structure
holds references to multiple fields within a collection's documents.
This allows the student to specify multiple text indexes without
violating the constraint for there to only be one text index per
collection. 

To enable text searches on the ``name`` and ``major`` field, the student
creates a compound index using the following snippet: 

.. code-block:: java

    collection.createIndex(Indexes.compoundIndex(Indexes.text("name"), Indexes.text("major")));

For more information, see the following documentation:

- :manual:`Text Indexes </core/index-text/>` Server Manual Entry
- :ref:`Text Indexes <index-text-indexes>` in our Indexes Builder Guide
- :ref:`Compound Indexes <index-compound-indexes>` in our Indexes Builder Guide

> Searching

Reviewer: For parallel structure with the "Text Index" heading
```suggestion
Text Search
```

.. code-block:: json
   :copyable: false

   { "_id": 2, "name": "Discrete Math in Computer Science", "major": ["Computer Science", "Mathematics"]}
   { "_id": 4, "name": "Computer science Seminar", "major": ["Computer Science"]}
 
Search Text with Terms Excluded
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Prefix each term you want to exclude in your text search with a minus
sign ``-``. 

Example 
```````

The student wants to find science courses that are not related to
computer science. 

To address this scenario, the student runs a text search on the
documents in the ``junior_courses`` collection that contain the term
"Science", but do not contain the term "Computer" in the ``name`` field.

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin negateExample
   :end-before: end negateExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 1, "name": "Forensic science", "major": ["Forensic Science", "Criminal Justice"] }
   { "_id": 3, "name": "Environmental Science", "major": ["Environmental Science", "Sustainable Management"] }

.. _sort_by_relevance:

> Sorting by Relevance

Reviewer: For parallel structure with the other sections
```suggestion
Sort by Relevance
```

~~~~~~~~~~~~~~~~~~~~

The text search assigns a numerical text score to indicate how relevant
each result matches the search string. Often, you want to know which
results are the most relevant.  

To sort by relevance, you use the :doc:`Projections
</fundamentals/builders/projections>`  and :doc:`Sorts
</fundamentals/builders/sort>` builders.

First, use the ``Projections.metaTextScore()`` method to retrieve a text
score. This method generates a new field with the name and returns the

> metadata associated with the text score.

Reviewer: Could be helpful to describe how to "use the Projections.metaTextScore() method", i.e. where does this fit into the query? 

I think the second sentence needs to be more specific about what generates a new field, what the "name" refers to, and what's in the "metadata associated with the text score" is. And is it associated with the text score or does it include the text score?

Author: I tried defining `metaTextScore` based on the definition here; https://docs.mongodb.com/manual/reference/operator/aggregation/meta/#mongodb-expression-exp.-meta

I simplified my definition and added more information for this section.

===========
Search Text
===========

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you will learn how to run a **text search** in MongoDB.

In MongoDB, you can use a text search to retrieve documents that
contain a term or a phrase in a specified field. A term is a sequence of
characters that excludes whitespace characters. A phrase is a sequence

> of terms that with any number of whitespace characters.

Reviewer: Nitpick:
I wonder if this is easier to read as separate paragraphs. Then you could list term and phrase with bold formatting so that the reader recognizes they're terms they should understand and remember.

Author: I tried it out and didn't think it was visually nice, so I kept it as it is.

Sample Documents
~~~~~~~~~~~~~~~~

In the following sections, you run text searches on the

> ``fast_and_furious_movies`` collection. Each section uses ``collection``

Reviewer: Could you specify "a variable named collection" rather than just monospace? The term "uses" is often indeterminate, so either we should add additional description around it or use a more descriptive word.

Reviewer: I think we use monospace for variable names rather than quotes. Can you update that?

to refer to the ``MongoCollection`` instance of the
``fast_and_furious_movies`` collection. 

This collection contains the following documents that represent one of
the several movies that are part of the Fast and Furious movie franchise
and tags they are mostly associated with:

.. code-block:: json

   { "_id": 1, "title": "2 Fast 2 Furious ", "tags": ["undercover", "drug dealer"] } 
   { "_id": 2, "title": "Fast 5", "tags": ["bank robbery", "full team"] }
   { "_id": 3, "title": "Furious 7", "tags": ["emotional"] }
   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }

Text Index
----------

You must create a **text index** before running a text search. A text
index specifies the string or string array field for the query to run a

> text search on. A collection can only contain one text index.

Reviewer: I don't think adding the preposition "on" at the end is grammatically correct because it doesn't relate the noun/verb to the preposition. Instead, consider "field on which to run a text search."

Reviewer: In the current version, I see:
"A text index specifies the string or string array field for the query to run a text search."

I don't think "to" is the right preposition. If the sentence was "To run a text search, do X", then "to" is appropriate. In this case, you should use "on which to run" as recommended in my prior comment because it refers to the field.

In the following examples, you run text searches on the ``title``
field in the ``fast_and_furious_movies`` collection. To enable text
searches on the ``title`` field, create a text index using the
:ref:`Indexes <index-text-indexes>` builder with the following 
snippet: 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin textIndex
   :end-before: end textIndex

> .. note::

Reviewer: I think specifying a text index on multiple fields should be in its own section and the note should only apply to identify that they need to drop the index. I think placing it in the note relates it only to indexes on multiple fields, but it's an operation we would want them to know for other use cases such as when they no longer wanted it or wanted to change it to a different field.
i

   If you want to run text searches on both the ``title`` and
   ``tags`` field in the ``fast_and_furious_movies`` collection, you
   need to create an text index to include them.

>    If you already have an existing text index, you need to drop the
>    index by specifying the existing text index field name with the
>    following snippet: 

Reviewer: ```suggestion
   If you have an existing text index on the collection, you need to drop
   it. You can drop an index by calling the ``dropIndex()`` method,
   specifying name of the text index as shown in the following snippet:
```

   .. code-block:: java

>       collection.dropIndex("field_name");

Reviewer: If you specify the process of dropping a text index by its field name, it's not clear to me how to drop a text index with multiple field names. Could you describe that somewhere? Or otherwise how to determine the name of the index?

Author: I mention in the sentence after you specify a *

Reviewer: That drops _all_ indexes including non-text indexes. I don't think this is a practical solution because you generally have other meaningful indexes on a collection to improve search/sort speed.

Author: I'm removing this part and redirecting it to the Indexes page.

>    If you want to delete all existing indexes except ``_id``, specify "*". 

Reviewer: ```suggestion
   If you want to delete all existing indexes except ``_id``, specify "*" instead of the name. 
```

   Then, you can create a :ref:`compound index <index-compound-indexes>`. A

>    compound index is a single index that holds references to multiple

Reviewer: I think "holds references" is jargon that is usually used when talking about pointers (reference to memory addresses). How about this:
```suggestion
   compound index is a single index that refers to multiple
```

   fields within a collection's documents. This allows you to specify
   multiple text fields using one text index.

   To enable text searches on the ``title`` and ``tag`` fields, create a
   compound index using the following snippet:  

   .. code-block:: java

      collection.createIndex(Indexes.compoundIndex(Indexes.text("title"), Indexes.text("tags")));

Text Search
-----------

Use the ``Filters.text()`` method to specify a text search. 

The ``Filters.text()`` method uses the :doc:`Filters builder
</fundamentals/builders/filters>` to create an instance of a :ref:`Bson
<bson>` type. The instance is a query filter that specifies how the text
search should operate, which you pass to the ``find()`` method. 

> When the ``find()`` method executes, MongoDB runs a text search on the

Reviewer: I think "when the X method executes" is passive voice. What is doing the execution? The subject of the dependent clause is not automatically related to any of the independent ones in the sentence.

fields indexed with the text index on the collection. The text search
returns documents that contain one or more of the search terms and a
relevance score for each result. For more information on relevance
scores, see the :ref:`Sort by Relevance <sort_by_relevance>` section. 

You can include ``TextSearchOptions`` as the second parameter of the
``Filters.text()`` method to specify text search options such as case

> sensitivity. By default, text searches run without case sensitivity

Reviewer: I think you're missing a description of what TextSearchOptions is in this revision. I thought you had a sentence about it before.

Author: It does say what it is in the first sentence. What are you looking for to be in the description?

Reviewer: Sorry, I must have been looking at an older version.

> which means MongoDB matches documents regardless of whether the value is
> lowercase or uppercase. 

Reviewer: Nitpick: just thought of a less wordy way to explain this.
```suggestion
which means the search matches lowercase and uppercase values.
```

To specify a case sensitive search, use the following snippet:

.. code-block:: java

   TextSearchOptions options = new TextSearchOptions().caseSensitive(true);
   Bson filter = Filters.text("SomeText", options);

The following sections show you how to perform the following types of
text searches:

- Search Text by a Term
- Search Text by a Phrase
- Search Text with Terms Excluded
- Sort by Relevance

For more information, see the following API documentation:

- :java-core-api:`Filters.text() <com/mongodb/client/model/Filters.html#text(java.lang.String)>`
- :java-core-api:`TextSearchOptions <com/mongodb/client/model/TextSearchOptions.html>`

.. _term_search:

Search Text by a Term
~~~~~~~~~~~~~~~~~~~~~

> Pass the term as a string to the ``Filters.text()`` method to specify a

Reviewer: ```suggestion
Pass a term as a string to the ``Filters.text()`` method to specify a
```

Reviewer: Bump ^

term in your text search. 

> If you want to match multiple terms in your text index, separate your
> terms with spaces. When you pass the ``Filters.text()`` method to the

Reviewer: ```suggestion
If you want to match multiple terms in your text search, separate them
with spaces. When you pass the ``Filters.text()`` method to the
```

> If you want to match multiple terms in your text index, separate your
> terms with spaces. When you pass the ``Filters.text()`` method to the
> ``find()`` method, the text search performs a logical OR on each term
> and returns documents that contain any of the terms. 

Reviewer: This one is tough to explain, so it might be best to introduce as few new concepts as possible. How about something like this:

```suggestion
If you want to match multiple terms in your text search, separate them
with spaces in the Filters.text builder method. When you pass the 
text search query returned by the builder method to the
``find()`` method, it returns documents that match any of the terms. 
```

Author: I'm removing this part and redirecting it to the Indexes page.

Reviewer: I still see this section which I think belongs here. It could help to identify the material with a header, perhaps something like "Search Text by Multiple Terms"

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the

> term "Fast".

Author: It's actually not a case sensitive search -- I forgot to change the name of the section

> term "Fast".

Reviewer: Ok cool. I wonder if you can then specify the search term as "fast" instead then as the capitalization doesn't change the result and could be confused a case-sensitive search.

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin caseSensitiveExample
   :end-before: end caseSensitiveExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 1, "title": "2 Fast 2 Furious ", "tags": ["undercover", "drug dealer"] }
   { "_id": 2, "title": "Fast 5", "tags": ["bank robbery", "full team"] }

Search Text by a Phrase
~~~~~~~~~~~~~~~~~~~~~~~

> Pass the phrase with escaped quotes to the ``Filters.text()`` method to

Reviewer: ```suggestion
Pass a phrase with escaped quotes to the ``Filters.text()`` method to
```

specify a phrase in your text search. If you don't add escaped quotes,
the method runs a :ref:`term search <term_search>`. 

Example 
```````

The following example runs a text search on the documents in the
``fast_and_furious_movies`` collection for titles that contain the
phrase "Fate of the Furious". 

.. literalinclude:: /includes/fundamentals/code-snippets/SearchText.java
   :language: java
   :dedent:
   :start-after: begin phraseExample
   :end-before: end phraseExample

The following shows the output of the preceding code:

.. code-block:: json
   :copyable: false

   { "_id": 4, "title": "The Fate of the Furious", "tags": ["betrayal"] }
 
Search Text with Terms Excluded
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Pass a term prefixed with a minus sign ``-`` to the ``Filters.text()``

> method to specify each term you want to exclude in your text search. 

Reviewer: I wonder if it might help to explain what it means to exclude a term. As a reader, I might confuse this with the following:
The results are the same, but the returned document comes back modified, with the excluded terms redacted

Reviewer: Just bumping this, not sure it was addressed.
One additional item is that the "-" doesn't stand out enough to tell that it's not just a dash in your text but rather an operator. Maybe you can surround it with parentheses or leave it out?

> 

Reviewer: I think it could be helpful to explain how exclusion works. For example, as a reader, I'm curious about the following situations:
1) Can I specify only an exclusion term without any inclusion terms?
2) If my text index includes multiple fields and one of those fields contains the exclusion term, does that mean it is not a match?

Example 
```````

The following example runs a text search on the documents in the

> ``fast_and_furious_movies`` collection for titles that contain the

Reviewer: Would it be helpful to mention what fields are part of the text index?

Author: It does say "for the titles". Where else do you think it should be specified?

Reviewer: I meant to mention what fields are part of the text index, not part of your search. I think you added a sentence somewhere near the beginning about what text index the examples use, so I guess omitting it is fine now.
